concurrent/src/main/java/io/airlift/concurrent/BoundedExecutor.java

```diff
@@ -52,35 +50,42 @@ public BoundedExecutor(Executor coreExecutor, int maxThreads)
     @Override
     public void execute(Runnable task)
     {
-        checkState(!failed.get(), "BoundedExecutor is in a failed state");
-
         queue.add(task);
 
         int size = queueSize.incrementAndGet();
         if (size <= maxThreads) {
-            // If able to grab a permit (aka size <= maxThreads), then we are short exactly one draining thread
             try {
                 coreExecutor.execute(this::drainQueue);
             }
             catch (Throwable e) {
-                failed.set(true);
-                log.error("BoundedExecutor state corrupted due to underlying executor failure");
+                decrementAndGetQueueSize();
```

Comment:
if `coreExecutor.execute(this::drainQueue)` throws immediately, you can end up having 1 task in queue and queueSize=0.

To me, it looks like you want something like "current thread count" instead of "queue size". Question is how to do this without losing an incoming task.